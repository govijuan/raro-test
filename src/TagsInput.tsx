import React, { useEffect } from "react";
import Chip from "@material-ui/core/Chip";
import TextField from "@material-ui/core/TextField";

import {validateEmail} from './helpers'

export default function TagsInput(props) {
  const { onSelectTags, tags, ...other } = props;
  const [ selectedItem, setSelectedItem ] = React.useState([]);


  useEffect(() => {
    setSelectedItem(tags);
  }, [tags]);

  const onKeyDown = e => {
    const key = e.code;
    switch(key){
      case 'Tab':
      case 'Enter':
        processTags(e);
        e.target.value = '';
        break;
      case 'Backspace':
        console.log("Backspace presionado!");
        handleBackspace(e);
        break;
      default:
        return;
    }
  }

  const processTags = e => {
    const value = e.target.value;
    let emailsForAdding=[]
    if(value.includes(';')){
      emailsForAdding = value.split(';').filter( item =>
                                                    item !== ''
                                                    && validateEmail(item) 
                                                    && !selectedItem.includes(item)
                                                );
      console.log("Split result: ", emailsForAdding);
    }else{
      if(validateEmail(value) && !selectedItem.includes(value)) emailsForAdding.push(value)
    }
    if(emailsForAdding.length !== 0) {
      setSelectedItem([...selectedItem, ...emailsForAdding]);
    }
  };

  const handleBackspace = (e) => {
    const newSelectedItem = [...selectedItem];
    newSelectedItem.pop();
    setSelectedItem(newSelectedItem);
  }

  const handleDelete = (item) => () => {
    const newTags = selectedItem.filter( setItem => setItem !== item);
    setSelectedItem(newTags);
  };

  return (
    <TextField
      InputProps={{
        startAdornment: selectedItem.map((item) => (
          <Chip
            key={item}
            tabIndex={-1}
            label={item}
            onDelete={handleDelete(item)}
            data-testid='email-tag'
          />
        ))
      }}
      {...other}
      onKeyDown={onKeyDown}
    />
  );
}
TagsInput.defaultProps = {
  tags: []
};
