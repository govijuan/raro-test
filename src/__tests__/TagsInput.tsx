import React from 'react'
import TagsInput from '@/TagsInput'
import { cleanup, render, fireEvent, screen, within } from '@testing-library/react'

describe('TagsInput Component', () => {
  afterEach(jest.clearAllMocks)
  afterEach(cleanup)

  const emailsArray = ['contato@rarolabs.com.br', 'nao-responda@rarolabs.com.br']

  const processKeyboardEvents = (keyboardEvt: string, emailsArr: any[]) => {
    render(<TagsInput placeholder='add Tags' tags={ keyboardEvt === 'Backspace' ? emailsArr : [] }/>)
    const input = screen.getByPlaceholderText('add Tags');
    if(keyboardEvt === 'Enter' || keyboardEvt === 'Tab') { 
      const inputValue = emailsArr.join(';')
      fireEvent.change(input, { target: { value:  inputValue} }) 
    }
    fireEvent.keyDown(input, { key: keyboardEvt, code: keyboardEvt })
    const inputTags = screen.getAllByTestId('email-tag')
    let expectedLength = 0
    switch(keyboardEvt) {
      case 'Enter':
      case 'Tab':
        expectedLength = emailsArr.length
        break;
      case 'Backspace':
        expectedLength = emailsArr.length - 1
        break;
    }
    expect(inputTags.length).toBe(expectedLength)
    const tagTexts = inputTags.map( tag => {
      const tagSpan = within(tag).getByText((content, element) =>{
        return element.tagName.toLocaleLowerCase() === 'span'
      })
      return tagSpan.textContent.toString()
    })
    if(keyboardEvt === 'Backspace'){
      expect(emailsArr).not.toEqual(tagTexts)
      expect(emailsArr).toEqual(expect.arrayContaining(tagTexts))
    } else {
      expect(emailsArr).toEqual(tagTexts)
    }
  }

  it('cria o snapshot do component', () => {
    const container = render(<TagsInput />)
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('deve renderizar as tags enviadas por atributos', () => {
    const emails = ['contato@rarolabs.com.br', 'nao-responda@rarolabs.com.br']

    const { debug } = render(<TagsInput tags={emails} />)

    debug()
  })

  it('deve renderizar tags quando preencher o input e pressionar enter', () => {
    processKeyboardEvents('Enter', emailsArray)
  })

  it('deve renderizar tags quando preencher o input e pressionar tab', () => {
    processKeyboardEvents('Tab', emailsArray)
  })

  it('deve deletar a útima tag criada ao pressionar o botão de backspace', async () => {
    processKeyboardEvents('Backspace', emailsArray )
  })
})
