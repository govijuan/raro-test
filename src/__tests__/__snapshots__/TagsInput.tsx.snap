// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsInput Component cria o snapshot do component 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-root-1"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd"
    >
      <div
        class="tags-container"
      />
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd"
        type="text"
        value=""
      />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Send mailing list
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
